azurerm/internal/services/apimanagement/api_management_custom_domain_resource.go

```diff
@@ -0,0 +1,325 @@
+			results = append(results, output)
+		}
+	}
+	if portalRawVal, ok := input.GetOk("portal"); ok {
+		vs := portalRawVal.([]interface{})
+		for _, rawVal := range vs {
+			v := rawVal.(map[string]interface{})
+			output := expandApiManagementCommonHostnameConfiguration(v, apimanagement.HostnameTypePortal)
+			results = append(results, output)
+		}
+	}
+	if developerPortalRawVal, ok := input.GetOk("developer_portal"); ok {
+		vs := developerPortalRawVal.([]interface{})
+		for _, rawVal := range vs {
+			v := rawVal.(map[string]interface{})
+			output := expandApiManagementCommonHostnameConfiguration(v, apimanagement.HostnameTypeDeveloperPortal)
+			results = append(results, output)
+		}
+	}
+	if proxyRawVal, ok := input.GetOk("proxy"); ok {
+		vs := proxyRawVal.([]interface{})
+		for _, rawVal := range vs {
+			v := rawVal.(map[string]interface{})
+			output := expandApiManagementCommonHostnameConfiguration(v, apimanagement.HostnameTypeProxy)
+			if value, ok := v["default_ssl_binding"]; ok {
+				output.DefaultSslBinding = utils.Bool(value.(bool))
+			}
+			results = append(results, output)
+		}
+	}
+	if scmRawVal, ok := input.GetOk("scm"); ok {
+		vs := scmRawVal.([]interface{})
+		for _, rawVal := range vs {
+			v := rawVal.(map[string]interface{})
+			output := expandApiManagementCommonHostnameConfiguration(v, apimanagement.HostnameTypeScm)
+			results = append(results, output)
+		}
+	}
+	return &results
+}
+
+func flattenApiManagementHostnameConfiguration(input *[]apimanagement.HostnameConfiguration, d *schema.ResourceData) []interface{} {
+	results := make([]interface{}, 0)
+	if input == nil {
+		return results
+	}
+
+	managementResults := make([]interface{}, 0)
+	portalResults := make([]interface{}, 0)
+	developerPortalResults := make([]interface{}, 0)
+	proxyResults := make([]interface{}, 0)
+	scmResults := make([]interface{}, 0)
+
+	for _, config := range *input {
+		output := make(map[string]interface{})
+
+		if config.HostName != nil {
+			output["host_name"] = *config.HostName
+		}
+
+		if config.NegotiateClientCertificate != nil {
+			output["negotiate_client_certificate"] = *config.NegotiateClientCertificate
+		}
+
+		if config.KeyVaultID != nil {
+			output["key_vault_id"] = *config.KeyVaultID
+		}
+
+		// Iterate through old state to find sensitive props not returned by API.
+		// This must be done in order to avoid state diffs.
+		// NOTE: this information won't be available during times like Import, so this is a best-effort.
+		snakeCaseConfigType := azure.ToSnakeCase(string(config.Type))
+		if valsRaw, ok := d.GetOk(snakeCaseConfigType); ok {
+			vals := valsRaw.([]interface{})
+			for _, val := range vals {
+				oldConfig := val.(map[string]interface{})
+
+				if oldConfig["host_name"] == *config.HostName {
+					output["certificate_password"] = oldConfig["certificate_password"]
+					output["certificate"] = oldConfig["certificate"]
+				}
+			}
+		}
```

Comment:
Since we're going to the trouble of switching on `config.Type` I think we should make this a helper func and move its use to the cases. It will involve a few more LoC, but will be more explicit about what we're attempting to get from config/state, and should help guard against changes to the structure in the API responses.